### PR TITLE
Use task-based rechunking to prechunk along partial boundaries

### PR DESCRIPTION
**Pull request by hendrikmakait**
This PR faciliates P2P rechunking by offloading pre-chunking along partial boundaries to task-based rechunking. This also functions as a predecessor to pre-concatenation of small chunks which can also be offloaded to P2P rechunking.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/shuffle/_rechunk.py [hunk 1]
```diff
@@ -204,32 +198,42 @@ class _NDPartial(NamedTuple):
...
-    #: partials representing the full rechunk.
     ix: NDIndex
 
 
 def rechunk_name(token: str) -> str:
     return f"rechunk-p2p-{token}"
 
 
-def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
+# FIXME: Pass these parameters on the dask/dask side
```
Comment (hendrikmakait) [inline distributed/shuffle/_rechunk.py:208]: TODO

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -64,7 +64,7 @@ repos:
           - tornado
           - pyarrow
           - urllib3
-          - git+https://github.com/dask/dask
+          - git+https://github.com/hendrikmakait/dask@prechunk-p2p-rechunk
```
Comment (hendrikmakait) [inline .pre-commit-config.yaml:67]: TODO: Revert before merging

File: distributed/shuffle/_rechunk.py [hunk 3]
```diff
@@ -431,52 +432,108 @@ def _construct_graph(self) -> _T_LowLevelGraph:
         return dsk
 
 
+def _prechunk_for_partials(
```
Comment (phofl) [inline distributed/shuffle/_rechunk.py:435]: The function name implies to me that we apply the rechunking here, could you rename to something that tells us that we are only calculating the chunks?

File: distributed/shuffle/tests/test_rechunk.py [hunk 4]
```diff
@@ -1350,4 +1351,39 @@ async def test_partial_rechunk_taskgroups(c, s):
...
+    assert len(s.task_groups) < 8
+
+
+@pytest.mark.parametrize(
+    ["old", "new", "expected"],
+    [
+        [((2, 2),), ((2, 2),), ((2, 2),)],
+        [((2, 2),), ((4,),), ((2, 2),)],
+        [((2, 2),), ((1, 1, 1, 1),), ((2, 2),)],
+        [((2, 2, 2),), ((1, 2, 2, 1),), ((1, 1, 1, 1, 2),)],
```
Comment (phofl) [inline distributed/shuffle/tests/test_rechunk.py:1363]: Why is the last 2 in expected a good idea?

Comment (phofl) [reply distributed/shuffle/tests/test_rechunk.py:1363]: Is it because the 1 in new at the end means that the input chunk only creates 2 output chunks?

**Review by phofl — APPROVED** [review]
only 2 nits, the logic seems to be fine that you added, I still don't quite understand the logic in ``_slice_new_chunks_into_partials`` but that't out of scope here

File: distributed/shuffle/tests/test_rechunk.py (hunk 4 above)
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_rechunk.py:1363]: The basic idea is that culling doesn't benefit from splitting the last two output chunks and there's "all-to-all" communication.

Comment (phofl) [reply distributed/shuffle/tests/test_rechunk.py:1363]: sounds good

Comment (hendrikmakait) [reply distributed/shuffle/tests/test_rechunk.py]: I've thought about this a little more, and you're right! There was a small bug here. I've adjusted the algorithm accordingly.

File: distributed/shuffle/_rechunk.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py]: Done

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
looks like there is an env mismatch of some kind because of tags

**Pull request merged**